core/server/worker/src/main/java/alluxio/worker/block/BlockLockManager.java

```diff
@@ -151,16 +186,22 @@ private OptionalLong lockBlockInternal(long sessionId, long blockId, BlockLockTy
     }
     try {
       long lockId = LOCK_ID_GEN.getAndIncrement();
-      try (LockResource r = new LockResource(mSharedMapsLock.writeLock())) {
-        mLockIdToRecordMap.put(lockId, new LockRecord(sessionId, blockId, lock));
-        Set<Long> sessionLockIds = mSessionIdToLockIdsMap.get(sessionId);
-        if (sessionLockIds == null) {
-          mSessionIdToLockIdsMap.put(sessionId, Sets.newHashSet(lockId));
-        } else {
-          sessionLockIds.add(lockId);
+      LockRecord record = new LockRecord(sessionId, blockId, lockId, lock);
+      if (mIsCleaningSessions.get()) {
+        synchronized (mSignal) {
+          while (mIsCleaningSessions.get()) {
+            mSignal.wait();
+          }
+          mLockRecords.add(record);
         }
+      } else {
+        mLockRecords.add(record);
```

Comment:
This might be a race condition🤦‍♂️. Even if `mIsCleaningSessions.get()` returns false in the if statement, the signal could be immediately turned into `true` before we proceed to add the record.